### PR TITLE
Fix documentation compilation

### PR DESCRIPTION
**Pull request by thierry-martinez**
@shinich1 noticed that documentation compilation was broken (see [comment](https://github.com/TeamGraphix/graphix/pull/198#issuecomment-2318179858)). This pull request fixes examples, cross-references and doc-comments to fix various errors. Another commit adds a `doc` workflow for CI validation of the documentation (as suggested by @EarlMilktea, https://github.com/TeamGraphix/graphix/pull/198#issuecomment-2318179858).

I used `sphinx-build --nitpicky` first to catch more errors, but I didn't manage to make it resolve references to `nx_reportviews.EdgeView`, so I removed the option.

### CODE REVIEW

**Comment by shinich1** [comment]
great PR, thanks!

**Comment by EarlMilktea** [comment]
@thierry-martinez 

Read the Docs has the official CI framework: https://docs.readthedocs.io/en/stable/integrations.html .
Why not use it?

**Comment by thierry-martinez** [comment]
We could configure Read the Docs CI, but the integration with GitHub seems to require admin rights on both sides (GitHub and Read the Docs). On the other hand, I was able to add the doc workflow without needing admin rights on either platform.

What are the benefits of using Read the Docs CI instead of GitHub Actions?

Perhaps we could merge this PR first since it fixes a number of errors and the CI check with GitHub Actions is working properly. We could then propose using Read the Docs CI in a separate issue or PR, where someone with admin rights can handle the setup. What do you think?

**Comment by EarlMilktea** [comment]
Either looks good to me.

**Comment by EarlMilktea** [comment]
Related to package documentation/comment, let me suggest an idea: why not use Ruff rule `D` for better documentation styling?
We may need to turn off `D10` for now to allow undocumented public items.

**Comment by shinich1** [comment]
> Perhaps we could merge this PR first since it fixes a number of errors and the CI check with GitHub Actions is working properly. We could then propose using Read the Docs CI in a separate issue or PR, where someone with admin rights can handle the setup. What do you think?

I think we can merge this PR after adding `D` rule as @EarlMilktea suggested. We should explore github pages which might integrate better with github actions.

**Review by shinich1 — APPROVED** [review]

**Comment by EarlMilktea** [comment]
Memo: May need to completely ignore rule `D` for `examples/*.py`.

**Comment by thierry-martinez** [comment]
I turned on the `D` rule in fd1c57c, and add/fixed the docstrings. `D` rule is ignored in `benchmarks`, `examples` and `tests`.

**Comment by EarlMilktea** [comment]
@thierry-martinez 

How thorough it is! Excellent!
Let me help you with finding typos...

**Comment by EarlMilktea** [comment]
@thierry-martinez 

Let me suggest some additional fixes on https://github.com/TeamGraphix/graphix/tree/fix-doc-ss .

**Comment by EarlMilktea** [comment]
MEMO: Opened #208 to speed up CI checks.

**Comment by thierry-martinez** [comment]
If there's anything else I should address before merging, please let me know!

File: .github/workflows/doc.yml
Comment (EarlMilktea) [inline .github/workflows/doc.yml:10]: Is there specific reasons to use 3.11?
If not, let's use 3.12 as it seems to be the fastest.

File: .github/workflows/doc.yml
Comment (EarlMilktea) [inline .github/workflows/doc.yml:9]:
```suggestion

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

```

File: docs/source/conf.py
Comment (EarlMilktea) [inline docs/source/conf.py:19]:
```suggestion
    "sphinx.ext.intersphinx",
```

File: docs/source/conf.py
Comment (EarlMilktea) [inline docs/source/conf.py:34]:
```suggestion
    "networkx": ("https://networkx.github.io/documentation/stable/", None),
```

**Review by EarlMilktea — APPROVED** [review]
Great! Only some minor suggestions.

File: .github/workflows/doc.yml
Comment (EarlMilktea) [inline .github/workflows/doc.yml:5]:
```suggestion
  pull_request:
    branches: ["master"]
  workflow_dispatch:
```

**Comment by shinich1** [comment]
@thierry-martinez  LGTM! please squash and merge after resolving conflict, if you think it's ready?

**Pull request merged**
